weejson/src/main/scala/com/rallyhealth/weejson/v1/BufferedValue.scala

```diff
@@ -162,6 +163,19 @@ object BufferedValue extends Transformer[BufferedValue] {
   case class Obj(value0: (String, BufferedValue)*) extends BufferedValue {
 
     override def toString: String = value0.map { case (k, v) => s""""$k": $v""" }.mkString("Obj(", ", ", ")")
+
+    override def equals(that: Any): Boolean = that match {
+      case Obj(thatValue0 @ _*) =>
+        this.value0.size == thatValue0.size &&
+          this.value0.sortBy(_._1).zip(thatValue0.sortBy(_._1)).forall {
+            case ((thisKey, thisValue), (thatKey, thatValue)) =>
+              thisKey == thatKey && thisValue == thatValue
+          }
+      case _ => super.equals(that)
+    }
+
+    // expensive but reliable
+    override def hashCode(): Int = this.value0.sortBy(_._1).hashCode()
```

Comment:
Why do we need an unordered equals/hashCode here? Can we have this do a cheap ordered comparison instead? Seems like that would be preferable anyway for use in tests.

Reply:
So you want `Obj("a" -> 1, "b" -> 2)` != `Obj("b" -> 2, "a" -> 1)`? Seems wrong. This goes back to the "dup key" conversation , and whether the underlying data structure for `Value`'s `Obj` (a `Map` which is unordered and does not allow duplicates) is superior to the underlying data structure for `BufferedValue` (a `Seq`, which is ordered and allows duplicates). You can argue against the underlying structure being optimal (and I think we agree on that point), but given what we have, this is the most appropriate definition for `equals` and `hashCode` IMO.

Reply:
I suppose one approach could be to do a cheap, unordered compare first, and only attempt the expensive, ordered compare if there is a mismatch. If transformations generally preserve key order and compared values generally match, compares would be cheaper. But if transformations frequently scramble key order or compared values generally mismatch, compares would be even more expensive (doing everything twice). I'm honestly not inclined to fine-tune this too much if our position is to just keep using `Value` for the time being, but it is important to me for it to at least be correct.

Reply:
Part of the reason this feels hard is that we lack a clear definition of what "correct equality" should mean for `BufferedValue`. Given its current role buffering input, how about:

> Two `BufferedValue` are equal if one can be substituted for another in a `bufferedValue.transform(to[T])` call for any valid `To[T]`, such that the resulting values of `T` are also equal.

This definition would clarify that losing precision over `Infinity` https://github.com/rallyhealth/weePickle/pull/107#discussion_r776868975 isn't valid. As for objects, I'm not sure, but I'm not yet convinced that we can eliminate the possibility of some data format or structure that permits and allows multimaps.

Reply:
That makes sense to me, and clearly both `Obj("a" -> 1, "b" -> 2)` and `Obj("b" -> 2, "a" -> 1)` would transform to the same `case class Thing(a: Int, b: Int)`, so we should consider them equal (somehow).

Reply:
Yeah, it works for case classes and scala Maps, but are we prepared to go so far as to say that a `To[T]` where the `T` cares about object order is invalid for all `To[T]`? Does this hold for `MongoDBObjects`? YAML? XML? The rest of the textual and binary formats in the jackson ecosystem and beyond?

Reply:
So now we agree on what "correct equality" means, but don't agree on what "valid" means in the context of "...for any valid To[T]" :'(

Did you mean "..for all valid To[T]"?